### PR TITLE
asim/kustoking/update-auditevent-microsoftwindows

### PR DESCRIPTION
**Pull request by KustoKing**
Required items, please complete
   
   Change(s):
   - Added new events

   Reason for Change(s):
   - Requested by customers

   Testing Completed:
   - Tested and added results in PR

   Checked that the validations are passing and have addressed any issues that are present:
   - Need to update values in datatester csv

### CODE REVIEW

**Review by devikamehra — CHANGES_REQUESTED** [review]

File: Parsers/ASimAuditEvent/Parsers/ASimAuditEventMicrosoftWindowsEvents.yaml [hunk 1]
```diff
@@ -24,70 +24,150 @@ ParserParams:
...
+    // Scheduled Task Event Ids
+    let ScheduledTaskEventIds = dynamic([4698,4699,4700,4701,4702]);
+    // Active Directory Replica Source Naming Context Event Ids
+    let ActiveDirectoryReplicaIds = dynamic([4929]);
+    // Firewall Event Ids
+    let FirewallEventIds = dynamic([5025,5027,5028,5029,5030,5034,5035,5037]);
+    // EventID Lookup
+    let EventIDLookup = datatable(EventID:int, Operation:string, EventType:string, Object:string, ObjectType:string)
+    [   
+        1102, "Delete Logs", "Delete", "Security Logs", "EventLog",
```
[the remark this answers is not in the document]
Comment (KustoKing) [reply Parsers/ASimAuditEvent/Parsers/ASimAuditEventMicrosoftWindowsEvents.yaml:40]: Done

File: Parsers/ASimAuditEvent/Parsers/ASimAuditEventMicrosoftWindowsEvents.yaml [hunk 2]
```diff
@@ -24,70 +24,150 @@ ParserParams:
...
+        4698, "Create Scheduled Task", "Create", "", "Scheduled Task",
+        4699, "Delete Scheduled Task", "Delete", "", "Scheduled Task",
+        4700, "Enable Scheduled Task", "Enable", "", "Scheduled Task",
+        4701, "Disable Scheduled Task ", "Disable", "", "Scheduled Task",
+        4702, "Update Scheduled Task", "Set", "", "Scheduled Task",
+        4929, "Remove Active Directory Replica Source Naming Context", "Delete", "", "Active Directory Replica Source Naming Context",
```
[the remark this answers is not in the document]
Comment (KustoKing) [reply Parsers/ASimAuditEvent/Parsers/ASimAuditEventMicrosoftWindowsEvents.yaml:46]: Done

File: Parsers/ASimAuditEvent/Parsers/ASimAuditEventMicrosoftWindowsEvents.yaml [hunk 3]
```diff
@@ -24,70 +24,150 @@ ParserParams:
...
+        5025, "Stop Firewall Service", "Disable", "Firewall Service", "Service",
+        5027, "Retrieve the Security Policy From The Local Storage", "", "Firewall Service", "Service",
+        5028, "Parse the new Security Policy", "", "Firewall Service", "Service",
```
[the remark this answers is not in the document]
Comment (KustoKing) [reply Parsers/ASimAuditEvent/Parsers/ASimAuditEventMicrosoftWindowsEvents.yaml:49]: Done

File: Parsers/ASimAuditEvent/Parsers/ASimAuditEventMicrosoftWindowsEvents.yaml [hunk 4]
```diff
@@ -24,70 +24,150 @@ ParserParams:
...
+        5029, "Initialize the Firewall Driver", "", "Firewall Service", "Service",
```
[the remark this answers is not in the document]
Comment (KustoKing) [reply Parsers/ASimAuditEvent/Parsers/ASimAuditEventMicrosoftWindowsEvents.yaml:50]: Done

File: Parsers/ASimAuditEvent/Parsers/ASimAuditEventMicrosoftWindowsEvents.yaml [hunk 5]
```diff
@@ -24,70 +24,150 @@ ParserParams:
...
+        5030, "Start the Firewall Service", "Start", "Firewall Service", "Service",
+        5034, "Stop Firewall Driver", "Stop", "Firewall Driver", "Driver",
+        5035, "Start Firewall Driver", "Start", "Firewall Driver", "Driver",
+        5037, "Terminating Firewall Driver", "", "Firewall Driver", "Driver"
```
[the remark this answers is not in the document]
Comment (KustoKing) [reply Parsers/ASimAuditEvent/Parsers/ASimAuditEventMicrosoftWindowsEvents.yaml:54]: Done

File: Parsers/ASimAuditEvent/Parsers/ASimAuditEventMicrosoftWindowsEvents.yaml [hunk 6]
```diff
@@ -24,70 +24,150 @@ ParserParams:
...
+    ];
+    let ParsedEvents = materialize (
```
[the remark this answers is not in the document]
Comment (KustoKing) [reply Parsers/ASimAuditEvent/Parsers/ASimAuditEventMicrosoftWindowsEvents.yaml:56]: Fixed

File: Parsers/ASimAuditEvent/Parsers/ASimAuditEventMicrosoftWindowsEvents.yaml [hunk 7]
```diff
@@ -24,70 +24,150 @@ ParserParams:
...
+    // Parse EventLog
+    let EventLog = ParsedEvents
+    | where EventID in(EventlogEventIds)
+    | project-away Task*, *DRA, SourceAddr;
+    // Parse Scheduled Task
+    let ScheduledTask = ParsedEvents
+    | where EventID in(ScheduledTaskEventIds)
+    | extend 
+        Object = TaskName,
+        NewValue = iff (
```
[the remark this answers is not in the document]
Comment (KustoKing) [reply Parsers/ASimAuditEvent/Parsers/ASimAuditEventMicrosoftWindowsEvents.yaml:120]: Fixed

File: Parsers/ASimAuditEvent/Parsers/ASimAuditEventMicrosoftWindowsEvents.yaml [hunk 8]
```diff
@@ -24,70 +24,150 @@ ParserParams:
...
+    // Parse ADR
+    let ActiveDirectoryReplica = ParsedEvents
+    | where EventID in(ActiveDirectoryReplicaIds)
+    | extend 
+        NewValue = SourceDRA,
+        OldValue = DestinationDRA,
+        SrcFQDN = SourceAddr
+    | extend 
+        Value = NewValue,
+        Object = OldValue
```
[the remark this answers is not in the document]
Comment (KustoKing) [reply Parsers/ASimAuditEvent/Parsers/ASimAuditEventMicrosoftWindowsEvents.yaml:137]: The old value is the object that has been deleted.

File: Parsers/ASimAuditEvent/Parsers/ASimAuditEventMicrosoftWindowsEvents.yaml [hunk 9]
```diff
@@ -24,70 +24,150 @@ ParserParams:
...
+    | where EventID in(FirewallEventIds)
+    | project-away Task*, *DRA, SourceAddr
+    ;
+    // Union Events
+    union EventLog, ScheduledTask, ActiveDirectoryReplica, WindowsFirewall
+        | project-rename 
+            ActorUserId = SubjectUserSid,
+            ActorSessionId = SubjectLogonId,
+            DvcId = _ResourceId,
+            DvcFQDN = Computer,
```
[the remark this answers is not in the document]
Comment (KustoKing) [reply Parsers/ASimAuditEvent/Parsers/ASimAuditEventMicrosoftWindowsEvents.yaml:151]: Done

File: Parsers/ASimAuditEvent/Parsers/ASimAuditEventMicrosoftWindowsEvents.yaml [hunk 10]
```diff
@@ -24,70 +24,150 @@ ParserParams:
...
+            ActingAppId = ClientProcessId,
```
[the remark this answers is not in the document]
Comment (KustoKing) [reply Parsers/ASimAuditEvent/Parsers/ASimAuditEventMicrosoftWindowsEvents.yaml:152]: Yes

File: Parsers/ASimAuditEvent/Parsers/ASimAuditEventMicrosoftWindowsEvents.yaml [hunk 11]
```diff
@@ -24,70 +24,150 @@ ParserParams:
...
+            EventEndTime= TimeGenerated,
+            EventProduct = 'Windows',
+            EventVendor = 'Microsoft',
+            EventSchemaVersion = '0.1.0',
+            EventSchema = 'AuditEvent',
+            EventOriginalType = tostring(EventID),
+            DvcIdType = iff (DvcId == "", "", "AzureResourceID"),
+            ActorUsername = iff (SubjectDomainName == "", SubjectUserName, strcat (SubjectDomainName, '\\', SubjectUserName)),
+            ActorUsernameType = iff (SubjectDomainName == "", 'Simple', 'Windows'),
+            ActorUserIdType = 'SID'
```
[the remark this answers is not in the document]
Comment (KustoKing) [reply Parsers/ASimAuditEvent/Parsers/ASimAuditEventMicrosoftWindowsEvents.yaml:167]: I would expect it, but added iff

File: Parsers/ASimAuditEvent/Parsers/vimAuditEventMicrosoftWindowsEvents.yaml [hunk 12]
```diff
@@ -50,122 +50,209 @@ ParserParams:
...
+                and (isnull(starttime) or TimeGenerated >= starttime) 
                 and (isnull(endtime) or TimeGenerated <= endtime)
-                and not(disabled)
-            | project EventID, EventData, _ResourceId, TimeGenerated, Computer, Type
-            | where EventID == 1102
+                and (array_length(srcipaddr_has_any_prefix) == 0)
+                and (array_length(actorusername_has_any) == 0 or EventData has_any (actorusername_has_any))
+                and (array_length(newvalue_has_any) == 0 or EventData has_any (newvalue_has_any))
+                and EventID in(FilteredParsedEventIds)
+                and (array_length(object_has_any) == 0 or EventData has_any(object_has_any))
```
Comment (devikamehra) [inline Parsers/ASimAuditEvent/Parsers/vimAuditEventMicrosoftWindowsEvents.yaml:112]: This filter will not work as expected. For example, for Object == "Security Logs", the text will not be part of the event and thus would get dropped even when it is required.

**Review by devikamehra — CHANGES_REQUESTED** [review]

File: Parsers/ASimAuditEvent/Parsers/ASimAuditEventMicrosoftWindowsEvents.yaml [hunk 13]
```diff
@@ -24,70 +24,151 @@ ParserParams:
...
+                TaskName:string,
+                TaskContent:string,
+                TaskContentNew:string,
+                ClientProcessId:string,
+                DestinationDRA:string,
+                SourceDRA:string,
+                SourceAddr:string
             ) 
-            with (regex=@'<(\w+)>([^<]*)<')
+        with (regex=@'<(\w+)>([^<]*)<')
```
[the remark this answers is not in the document]
Comment (KustoKing) [reply Parsers/ASimAuditEvent/Parsers/ASimAuditEventMicrosoftWindowsEvents.yaml:86]: Done

File: Parsers/ASimAuditEvent/Parsers/vimAuditEventMicrosoftWindowsEvents.yaml [hunk 14]
```diff
@@ -50,122 +50,209 @@ ParserParams:
...
             SecurityEvent
-            | where 
-                (isnull(starttime) or TimeGenerated >= starttime) 
+            | where not(disabled)
+                and (isnull(starttime) or TimeGenerated >= starttime) 
                 and (isnull(endtime) or TimeGenerated <= endtime)
-                and not(disabled)
-            | project EventID, EventData, _ResourceId, TimeGenerated, Computer, Type
-            | where EventID == 1102
+                and (array_length(srcipaddr_has_any_prefix) == 0)
```
[the remark this answers is not in the document]
Comment (KustoKing) [reply Parsers/ASimAuditEvent/Parsers/vimAuditEventMicrosoftWindowsEvents.yaml:108]: Done

File: Parsers/ASimAuditEvent/Parsers/vimAuditEventMicrosoftWindowsEvents.yaml [hunk 15]
```diff
@@ -50,122 +50,209 @@ ParserParams:
...
-      eventtype_in = eventtype_in,
-      eventresult = eventresult,
-      operation_has_any = operation_has_any,
-      object_has_any=object_has_any,
-      newvalue_has_any=newvalue_has_any,
-      disabled=disabled
-   )
\ No newline at end of file
+        Object = TaskName,
+        NewValue = coalesce(
```
[the remark this answers is not in the document]
Comment (KustoKing) [reply Parsers/ASimAuditEvent/Parsers/vimAuditEventMicrosoftWindowsEvents.yaml:191]: Done

File: Parsers/ASimAuditEvent/Parsers/vimAuditEventMicrosoftWindowsEvents.yaml (hunk 12 above)
Comment (KustoKing) [reply Parsers/ASimAuditEvent/Parsers/vimAuditEventMicrosoftWindowsEvents.yaml:112]: Done

File: Parsers/ASimAuditEvent/Parsers/vimAuditEventMicrosoftWindowsEvents.yaml [hunk 16]
```diff
@@ -50,122 +50,209 @@ ParserParams:
...
-        | where EventID == 1102
-        | where (array_length(srcipaddr_has_any_prefix) == 0)
-           and (array_length(newvalue_has_any) == 0)
-           and (array_length(object_has_any) == 0 or "Logs" in (object_has_any))
-           and (array_length(operation_has_any) == 0 or "Delete Logs" has_any (operation_has_any))
-           and (array_length(eventtype_in) == 0 or "Delete" in (eventtype_in))
+            and (array_length(srcipaddr_has_any_prefix) == 0)
+            and (array_length(actorusername_has_any) == 0 or EventData has_any (actorusername_has_any))   
+            and (array_length(newvalue_has_any) == 0 or EventData has_any (newvalue_has_any))
+            and EventID in(FilteredParsedEventIds)
```
[the remark this answers is not in the document]
Comment (KustoKing) [reply Parsers/ASimAuditEvent/Parsers/vimAuditEventMicrosoftWindowsEvents.yaml]: Done

File: Parsers/ASimAuditEvent/Parsers/ASimAuditEventMicrosoftWindowsEvents.yaml [hunk 17]
```diff
@@ -24,70 +24,180 @@ ParserParams:
...
+        5029, "Initialize the Firewall Driver", "Initialize", "Firewall Service", "Service", "Failure",
+        5030, "Start the Firewall Service", "Start", "Firewall Service", "Service", "Failure",
+        5034, "Stop Firewall Driver", "Stop", "Firewall Driver", "Driver", "Failure",
+        5035, "Start Firewall Driver", "Start", "Firewall Driver", "Driver", "Failure",
+        5037, "Terminating Firewall Driver", "Terminate", "Firewall Driver", "Driver", "Failure",
+        7035, "Start Control Sent", "Execute", "", "Service", "Success",
+        7036, "Enter Stop State", "Stop", "", "Service", "Success",
+        7040, "Changed Service Settings", "Set", "", "Service", "Success",
+        7045, "Install Service", "Install", "", "Service", "Success",
+        2009, "Load Group Policy", "Error", "", "Service", "Failure",
```
Comment (devikamehra) [inline Parsers/ASimAuditEvent/Parsers/ASimAuditEventMicrosoftWindowsEvents.yaml:63]: Error is not a valid EventType. We can rather use Other in this case.

File: Parsers/ASimAuditEvent/Tests/Microsoft_Windows_AuditEvent_DataTest.csv [hunk 18]
```diff
@@ -0,0 +1,10 @@
+﻿Result
+"(0) Error: 3 invalid value(s) (up to 10 listed) in 249 records (100.0%) for field [ObjectType] of type [Enumerated]: [""Active Directory Replica Source Naming Context"",""Scheduled Task"",""EventLog""] (Schema:AuditEvent)"
```
Comment (devikamehra) [inline Parsers/ASimAuditEvent/Tests/Microsoft_Windows_AuditEvent_DataTest.csv:2]: Can you please add the latest test results.

File: Parsers/ASimWebSession/Parsers/ASimWebSessionFortinetFortiGate.yaml [hunk 19]
```diff
@@ -0,0 +1,126 @@
+Parser:
```
Comment (devikamehra) [inline Parsers/ASimWebSession/Parsers/ASimWebSessionFortinetFortiGate.yaml:1]: Please raise a separate PR for Fortinet parser.

**Review by devikamehra — CHANGES_REQUESTED** [review]

File: Parsers/ASimAuditEvent/Parsers/ASimAuditEventMicrosoftWindowsEvents.yaml [hunk 20]
```diff
@@ -24,70 +24,180 @@ ParserParams:
...
+        4929, "Remove Active Directory Replica Source Naming Context", "Delete", "", "Other", "Success",
+        5025, "Stop Firewall Service", "Disable", "Firewall Service", "Service", "Success",
+        5027, "Retrieve the Security Policy From The Local Storage", "Read", "Firewall Service", "Service", "Failure",
+        5028, "Parse the new Security Policy", "Set", "Firewall Service", "Service", "Failure",
+        5029, "Initialize the Firewall Driver", "Initialize", "Firewall Service", "Service", "Failure",
+        5030, "Start the Firewall Service", "Start", "Firewall Service", "Service", "Failure",
+        5034, "Stop Firewall Driver", "Stop", "Firewall Driver", "Driver", "Failure",
+        5035, "Start Firewall Driver", "Start", "Firewall Driver", "Driver", "Failure",
+        5037, "Terminating Firewall Driver", "Terminate", "Firewall Driver", "Driver", "Failure",
+        7035, "Start Control Sent", "Execute", Service "Service", "Success",
```
Comment (devikamehra) [inline Parsers/ASimAuditEvent/Parsers/ASimAuditEventMicrosoftWindowsEvents.yaml:59]: Value should be in double quotes. This is causing the KQL Validation failure.

**Review by devikamehra — CHANGES_REQUESTED** [review]

File: Parsers/ASimAuditEvent/Parsers/ASimAuditEventMicrosoftWindowsEvents.yaml (hunk 20 above)
Comment (KustoKing) [reply Parsers/ASimAuditEvent/Parsers/ASimAuditEventMicrosoftWindowsEvents.yaml]: done

File: Sample Data/ASIM/Microsoft_Windows_AuditEvent_IngestedLogs.csv [hunk 21]
```diff
@@ -0,0 +1,2682 @@
...
+"00000000-0000-0000-0000-000000000000","4/29/2023, 1:07:14.806 PM",OpsManager,"KUSTOWORKS\DC1$",Machine,"DC1.kustoworks.com","Microsoft-Windows-Security-Auditing",Security,1,0,"<EventData xmlns=""http://schemas.microsoft.com/win/2004/08/events/event"">
```
Comment (devikamehra) [inline Sample Data/ASIM/Microsoft_Windows_AuditEvent_IngestedLogs.csv:84]: Schema of this csv will not be able to accommodate both Security and Windows Event format. Please raise a separate file for both of them

File: Parsers/ASimAuditEvent/Tests/Microsoft_Windows_AuditEvent_DataTest.csv [hunk 22]
```diff
@@ -0,0 +1,12 @@
+﻿Result
+"(0) Error: 1 invalid value(s) (up to 10 listed) in 1541 records (78.54%) for field [EventType] of type [Enumerated]: [""Stop""] (Schema:AuditEvent)"
+"(0) Error: 4 invalid value(s) (up to 10 listed) in 1960 records (99.9%) for field [ObjectType] of type [Enumerated]: [""Event Log"",""Scheduled Task"",""Service"",""Directory Service Object""] (Schema:AuditEvent)"
+"(2) Info: Empty value in 1592 records (81.14%)  in recommended field [Object] (Schema:AuditEvent)"
```
Comment (devikamehra) [inline Parsers/ASimAuditEvent/Tests/Microsoft_Windows_AuditEvent_DataTest.csv:4]: Is there a specific reason why Object is going empty. According to the schema this value mandatory.

**Review by devikamehra — CHANGES_REQUESTED** [review]

File: Sample Data/ASIM/Microsoft_Windows_AuditEvent_IngestedLogs.csv (hunk 21 above)
Comment (KustoKing) [reply Sample Data/ASIM/Microsoft_Windows_AuditEvent_IngestedLogs.csv:84]: Sorted also for Event

File: Parsers/ASimAuditEvent/Tests/Microsoft_Windows_AuditEvent_DataTest.csv (hunk 22 above)
Comment (KustoKing) [reply Parsers/ASimAuditEvent/Tests/Microsoft_Windows_AuditEvent_DataTest.csv:4]: I modified the parser after the tests

**Review by devikamehra — APPROVED** [review]
Both parsers are tested

**Review by v-atulyadav — APPROVED** [review]

**Pull request merged**